plugins/out_s3/s3.c

```diff
@@ -1073,6 +1104,10 @@ static int upload_data(struct flb_s3 *ctx, struct s3_file *chunk,
         if (chunk) {
             s3_store_file_unlock(chunk);
             chunk->failures += 1;
+            if (chunk->failures > ctx->ins->retry_limit){
+                less_than_limit = FLB_FALSE;
+            }
+            s3_retry_warn(ctx, (char *)chunk->fsf->meta_buf, m_upload->input_name, chunk->create_time, less_than_limit);
```

Comment:
Sorry I think I missed this before, I think we need a change in the logic here. If we are over the limit, we should not unlock the chunk and should delete it instead and should then return FLB_ERROR instead of FLB_RETRY.

I see that we have this in the `put_all_chunks` case:

```
if (chunk->failures >= ctx->ins->retry_limit) {
                s3_retry_warn(ctx, (char *) fsf->meta_buf, NULL, chunk->create_time, FLB_FALSE);
                flb_fstore_file_inactive(ctx->fs, fsf);
```

So in the normal upload case, I think we need to call `flb_fstore_file_inactive` as well to remove the failed chunk,